### ISSUE TITLE
TileLang-Ascend Practice Assignment

### ISSUE THREAD

**Issue by fuhouyu-hw** [tile-ai/tilelang-ascend#431]
# TileLang-Ascend Practice Assignment

### The Bottleneck of Long Reasoning
Modern Large Language Models (LLMs), particularly those designed for complex reasoning (e.g., Chain-of-Thought), often generate extremely long sequences. While longer contexts correlate with better reasoning performance, they hit a critical bottleneck: standard attention mechanisms scale quadratically $O(N^2)$ in computation and require linear memory growth for Key-Value (KV) caches. This results in high latency and excessive memory I/O on hardware accelerators.

### The Sparsity Observation
Recent research observes that attention in reasoning tasks is **intrinsically sparse**. During decoding, a token typically attends to only a small subset of previous information. Crucially, this sparsity often exhibits locality, meaning relevant information tends to cluster in contiguous chunks rather than being randomly scattered.

### The Block Sparse Solution
**Block Sparse Attention** leverages this observation by partitioning the context into fixed-size blocks (e.g., $64 \times 64$). By utilizing a **block mask**—derived from heuristics or lightweight gating networks—the hardware can skip computation for irrelevant blocks entirely. This approach offers a dual advantage:
1.  **Arithmetic Efficiency:** It bypasses the massive floating-point operations required for negligible attention scores.
2.  **Memory Bandwidth Optimization:** It drastically reduces the data movement required to load KV blocks from High Bandwidth Memory (HBM) to on-chip SRAM.

In this challenge, you will implement a **Block Sparse Flash Attention** kernel on the **Huawei Ascend NPU** using TileLang. You will translate a mask-based skipping logic into NPU-compatible instructions (like `T.copy` and `T.gemm_v0`), optimizing for the Ascend architecture's specific memory hierarchy (Global Memory $\to$ L1 $\to$ L0/Unified Buffer).

![Image](https://github.com/user-attachments/assets/8926f65a-8f63-4d35-ab78-b979efb7c71f)

## Part 2: Problem Description

**block_sparse_flash_attention**

**`block_sparse_flash_attention(query, key, value, block_mask, block_size=64, is_causal=True) -> Tensor`**

Computes the block-sparse scaled dot-product attention independently over the Batch ($B$) and Head ($H$) dimensions. This operation combines the memory efficiency of FlashAttention with a user-defined block-level sparsity mask and a causal constraint.

For input tensors $Q, K, V \in \mathbb{R}^{B \times H \times S \times D}$ and a boolean block mask $BM \in \{0, 1\}^{B \times H \times S' \times S'}$ (where $S' = \lceil S / \text{blockSize} \rceil$), the output $O \in \mathbb{R}^{B \times H \times S \times D}$ is computed as follows.

For each batch index $b \in [0, B-1]$ and head index $h \in [0, H-1]$:

$$
O_{b,h} = \text{softmax}\left(\frac{Q_{b,h} (K_{b,h})^T}{\sqrt{D}} + M_{b,h}\right) V_{b,h}
$$

Where $Q_{b,h}, K_{b,h}, V_{b,h}$ represent the $S \times D$ matrices for the specific batch and head. The bias term $M_{b,h} \in \mathbb{R}^{S \times S}$ incorporates both the block-sparse mask and the causal mask. For any pair of sequence indices $i, j \in [0, S-1]$ (where $i$ is the query index and $j$ is the key index):

$$
M_{b,h}[i, j] = 
\begin{cases} 
0 & \text{if } \text{MaskCondition}(b, h, i, j) \text{ is True} \\
-\infty & \text{otherwise}
\end{cases}
$$

The **MaskCondition** is true if and only if **both** of the following hold:
**Block Validity:** The corresponding block in the `block_mask` is active (True).
`BM[b, h, floor(i/block_size), floor(j/block_size)] = 1`

**Causal Constraint:** The query index is greater than or equal to the key index (if `is_causal` is enabled).
`i ≥ j`

**Parameters:**

*   **query** (Tensor): Input tensor of shape $(B, H, S, D)$.
*   **key** (Tensor): Input tensor of shape $(B, H, S, D)$.
*   **value** (Tensor): Input tensor of shape $(B, H, S, D)$.
*   **block_mask** (Tensor): A coarse-grained boolean mask of shape $(B, H, S', S')$, where $S' = \lceil S / \text{blockSize} \rceil$. This determines strictly which blocks participate in the attention calculation.
*   **block_size** (int, optional): The tile size used for sparsity granularity. Default: 64.
*   **is_causal** (bool, optional): If `True`, applies a lower-triangular mask ($i \geq j$) on top of the block mask. Default: `True`.

**Output:**

*   **output** (Tensor): The computed attention result of shape $(B, H, S, D)$.

![Image](https://github.com/user-attachments/assets/794ca6b3-a0d5-48cb-9951-daefc65b765e)

## Part 3: Reference Implementation Method

The reference implementation adapts the standard FlashAttention-2 algorithm to the NPU architecture by introducing a block-sparse check. Unlike the GPU implementation which relies on shared memory and thread-level parallelism, the NPU implementation explicitly manages data movement across the **Global Memory (HBM)**, **L1 Buffer**, and the specialized computing units: **Cube Core** (Matrix Multiply) and **Vector Core** (Element-wise ops).

The key optimization is leveraging the coarse-grained mask to skip entire blocks of computation. If a $64 \times 64$ block is masked out, we bypass loading $K/V$ blocks and the associated matrix multiplications completely.

**Algorithm 1: Block-Sparse FlashAttention Forward Pass (NPU)**

**Require:** Matrices $\mathbf{Q}, \mathbf{K}, \mathbf{V} \in \mathbb{R}^{N \times d}$ in HBM.
**Require:** Block Mask $\mathbf{M} \in \mathbb{R}^{T_r \times T_c}$ in HBM, where $T_r = \lceil N/B_r \rceil, T_c = \lceil N/B_c \rceil$.
**Require:** Block sizes $B_r=64, B_c=64$.

1.  **Partition**: Divide $\mathbf{Q}$ into $T_r$ blocks, and $\mathbf{K}, \mathbf{V}$ into $T_c$ blocks.
2.  **Outer Loop** ($i = 0$ to $T_r - 1$):
    1.  **Load** $\mathbf{Q}_i$ from HBM to **L1 Buffer**.
    2.  **Initialize** accumulators on chip:
        *   $\mathbf{O}_i = \mathbf{0} \in \mathbb{R}^{B_r \times d}$ (in **L0C Buffer**)
        *   $\ell_i = \mathbf{0}, m_i = -\infty \in \mathbb{R}^{B_r}$ (in **Unified Buffer - UB**)
    3.  **Inner Loop** ($j = 0$ to $T_c - 1$):
        1.  **Check Mask**: $mask_{val} \leftarrow \mathbf{M}[batch, head, i, j]$.
        2.  **If** $mask_{val} == 0$: **Continue** (Skip computation).
        3.  **Else**:
            *   **Load** $\mathbf{K}_j$ from HBM to **L1 Buffer**.
            *   **[Cube Unit]** Compute scores $\mathbf{S}_{ij} = \mathbf{Q}_i \mathbf{K}_j^T$ into **L0C Buffer**.
            *   **Transfer** $\mathbf{S}_{ij}$ from **L0C** to **UB**.
            *   **[Vector Unit]** Update Softmax statistics in **UB**:
                *   $m_{new} = \max(m_i, \text{rowmax}(\mathbf{S}_{ij}))$
                *   Compute `P_ij = exp(S_ij - m_new)`
                *   $\ell_{new} = \ell_i \cdot e^{m_i - m_{new}} + \text{rowsum}(\mathbf{P}_{ij})$
                *   *Update accumulators:* $m_i \leftarrow m_{new}, \ell_i \leftarrow \ell_{new}$.
            *   **Transfer** $\mathbf{P}_{ij}$ (cast to fp16) from **UB** to **L1 Buffer**.
            *   **Load** $\mathbf{V}_j$ from HBM to **L1 Buffer**.
            *   **[Cube Unit]** Accumulate O_i = O_i * exp(scale) + P_ij * V_j in **L0C Buffer**.
    4.  **[Vector Unit]** Finalize $\mathbf{O}_i$ in **UB**: $\mathbf{O}_i = \mathbf{O}_i / \ell_i$.
    5.  **Write** $\mathbf{O}_i$ from **UB** to HBM.
3.  **Return** $\mathbf{O}$.

## Part 4: Code Template

```python
import math
import time
import torch
import torch_npu
import tilelang
import tilelang.language as T
import torch.nn.functional as F

tilelang.disable_cache()

def get_sparse_attn_mask_from_topk(x, topk, use_dense_for_last_block=False):
    bsz, num_head, downsample_len, _ = x.shape
    sparse_index = torch.topk(x, topk, dim=-1).indices
    dense_mask = torch.full([bsz, num_head, downsample_len, downsample_len], False, dtype=torch.bool, device=x.device)
    dense_mask.scatter_(-1, sparse_index, True)
    if use_dense_for_last_block:
        dense_mask[:, :, -2:, :] = True
    dense_mask.tril_()
    return dense_mask

@tilelang.jit(out_idx=[4])
def blocksparse_flashattn_ascend(batch, heads, seq_len, dim, downsample_len):
    block_M, block_N = 64, 64
    assert seq_len % block_M == 0
    
    shape = [batch, heads, seq_len, dim]
    block_mask_shape = [batch, heads, downsample_len, downsample_len]
    dtype, block_mask_dtype = "float16", "int8"
    total_blocks = (seq_len // block_M) * heads * batch

    @T.prim_func
    def main(
        Q: T.Tensor(shape, dtype),
        K: T.Tensor(shape, dtype),
        V: T.Tensor(shape, dtype),
        BlockSparseMask: T.Tensor(block_mask_shape, block_mask_dtype),
        Output: T.Tensor(shape, dtype),
    ):
        with T.Kernel(total_blocks, is_npu=True) as (cid, vid):
            blocks_per_seq = seq_len // block_M
            bx = cid % blocks_per_seq
            remaining = cid // blocks_per_seq
            by = remaining % heads
            bz = remaining // heads
            
            # TODO: Allocate buffers for Q, K, V, Output, and Accumulator
            q_ub = T.alloc_shared([block_M, dim], dtype)
            
            # TODO: Implement Cross-Core Pipeline using T.Pipelined
            
            # TODO: Implement Block Sparse Mask Loading
            
            # TODO: Implement GEMM and Softmax loop
            
            # TODO: Store results
            
            T.copy(Q[bz, by, bx * block_M : (bx + 1) * block_M, :], q_ub)
            T.barrier_all()
            T.copy(q_ub, Output[bz, by, bx * block_M : (bx + 1) * block_M, :])

    return main

def run_tests():
    torch.set_default_device('npu')
    torch.manual_seed(0)
    
    print("[Correctness Test]")
    B, H, S, D, TOPK, BLK = 1, 1, 256, 64, 2, 64
    
    q = torch.randn(B, H, S, D, dtype=torch.float16)
    k = torch.randn(B, H, S, D, dtype=torch.float16)
    v = torch.randn(B, H, S, D, dtype=torch.float16)
    
    ds_len = math.ceil(S / BLK)
    x_ds = torch.randn([B, H, ds_len, ds_len], dtype=torch.float16)
    x_ds[:, :, :, 0] = 100 
    mask = get_sparse_attn_mask_from_topk(x_ds, TOPK).to(torch.int8)
    
    kernel = blocksparse_flashattn_ascend(B, H, S, D, ds_len)
    out = kernel(q, k, v, mask)
    torch.npu.synchronize()
    
    q_cpu, k_cpu, v_cpu = q.cpu().float(), k.cpu().float(), v.cpu().float()
    mask_cpu = mask.cpu()
    
    ones_cpu = torch.ones(BLK, BLK, device='cpu')
    full_mask = torch.kron(mask_cpu.float(), ones_cpu)
    full_mask = full_mask[..., :S, :S].bool() & torch.tril(torch.ones(S, S, device='cpu')).bool()
    
    attn = torch.einsum("bhsd,bhtd->bhst", q_cpu, k_cpu) * (D**-0.5)
    attn = attn.masked_fill(~full_mask, float("-inf")).softmax(dim=-1)
    ref = torch.einsum("bhst,bhtd->bhsd", attn, v_cpu)
    
    try:
        torch.testing.assert_close(out.cpu().float(), ref, atol=1e-2, rtol=1e-2)
        print("Test Passed!")
    except AssertionError:
        print("Result mismatch (Expected before implementation)")

if __name__ == "__main__":
    run_tests()
```

## Part 5: Additional Guidance

This assignment is a practical exercise for the "TileLang Developer Programming Paradigm Detailed Explanation" course. As such, it is recommended to utilize Developer development mode features. You can enable various development features through the pass_configs parameter in the @tilelang.jit decorator. These features include, but are not limited to: TL_ASCEND_AUTO_CV_COMBINE, TL_ASCEND_AUTO_CV_SYNC, TL_ASCEND_MEMORY_PLANNING, and TL_ASCEND_AUTO_SYNC.

Additionally, you can leverage TileLang primitives such as T.Pipelined to implement double buffering and optimize data movement between different memory hierarchies. You may also use the msprof tool to measure performance.

Evaluation Process: We will use a unified testing environment to evaluate all submitted Pull Requests (PRs). The PR with the highest score and best performance will be selected for merging into the code repository.
